### PR TITLE
Feature/validate cli

### PR DESCRIPTION
**Pull request by sametd**
## Summary                                                                                                                                         
                                                                                                                                                     
  New `validate` subcommand and `validate_message()` / `validate_file()` library API for checking `.tgm` files are well-formed and intact — analogous
   to `grib_check` or `h5check`. This is **PR 1 of 3** for the tensogram-validate feature.                                                           
                                                                                                                                                     
  Three validation levels:                                                                                                                           
   
  - **Level 1 — Structure**: magic bytes, frame headers/ENDF, total_length, postamble, first_footer_offset, frame ordering, preceder legality,       
  preamble flags consistency                                                                                                                         
  - **Level 2 — Metadata**: CBOR parses correctly, required keys present, dtype/encoding/filter/compression recognized, shape/strides/ndim           
  consistent, index offsets match actual frame positions                                                                                             
  - **Level 3 — Integrity**: xxh3 hash verification (descriptor + hash frame fallback), compressed payloads decompress without error                 
                                                                                                                                                     
  ### CLI                                                                                                                                            
                                                                                                                                                     
  ```bash                                                                                                                                            
  tensogram validate [--quick|--checksum|--canonical] [--json] <FILES>...                                                                            
                                                                                                                                                     
  - default — Levels 1–3                                                                                                                             
  - --quick — Level 1 only, no payload access                                                                                                        
  - --checksum — Hash verification (structural errors still reported)                                                                                
  - --canonical — Levels 1–3 plus RFC 8949 CBOR key ordering check                                                                                   
  - --json — Machine-parseable JSON array output via serde_json                                                                                      
  - Mutually exclusive mode flags, batch mode, exit code 0/1                                                                                         
 ```                                                                                                                                                    
  **Architecture**                                                                                                                                       
                                                                                                                                                     
  Modular layout in crates/tensogram-core/src/validate/:                                                                                             
                                                                                                                                                     
  - types.rs — public types, ~40 stable IssueCode variants with serde serialization                                                                  
  - structure.rs — Level 1 raw byte walking with overflow-safe arithmetic                                                                            
  - metadata.rs — Level 2 CBOR parsing on raw frame payloads (before decode_message normalization)
  - integrity.rs — Level 3 hash verification and decode pipeline execution                                                                           
  - mod.rs — public API entry points and tests                    
                                                                                                                                                     
  **Key design decisions**                                                                                                                               
                                                                                                                                                     
  - Streaming file validation: validate_file() uses scan_file + per-message reads — O(1 message) memory, not O(file size)                            
  - Raw frame access for Level 2: parses CBOR from raw frame payloads, not through decode_message() which normalizes (resizes base array, strips     
  _reserved_ from preceders)                                                                                                                         
  - CBOR-before layout: uses Cursor position to measure consumed CBOR bytes (not reserialization)                                                    
  - Checksum mode: suppresses structural warnings but surfaces structural errors
  - Hash verification: hash_verified is true only when ALL objects are actually verified; unknown hash algorithms produce warnings                   
  - Missing hashes: reported as warnings (not errors) — the wire format says hashes are optional
  - Index frame: optional — not required for validation                                                                                              
  - Canonical CBOR: opt-in via --canonical, reported as warnings                                                                                     
  - Stable issue codes: machine-readable IssueCode enum (e.g. hash_mismatch, invalid_magic) serialized as snake_case strings                         
                                                                                                                                                     
  **What's next**                                                                                                                                        
                                                                                                                                                     
  - PR 2: Level 4 Fidelity (--full flag) — full decode, NaN/Inf warnings, encode.rs test cleanup                                                     
  - PR 3: Python bindings, C FFI, examples                        
                                                                                                                                                     
  **Test plan**                                                                                                                                          
                                         
  - cargo fmt — clean                                                                                                                                
  - cargo clippy --workspace --all-targets --all-features -- -D warnings — clean
  - cargo test --workspace — 670 tests pass, 0 failures
  - 5 golden .tgm files pass validation                                                                                                              
  - JSON output verified with Python json.load()
  - 52 validate-specific tests covering:                                                                                                             
    - Valid messages (single, multi-object, streaming, zero-object)
    - Structure failures (wrong magic, truncated, bad total_length, corrupted postamble)                                                             
    - Metadata failures (corrupted CBOR)                                                                                                             
    - Integrity failures (corrupted payload, missing hash)                                                                                           
    - All modes (quick, checksum, canonical, default)                                                                                                
    - Checksum mode on broken messages (structural errors surfaced)                                                                                  
    - File-level validation (trailing garbage, gaps between messages, truncated second message, garbage-only buffer, empty buffer)                   
    - Streaming messages with footer metadata                                                                                                        
    - Corrupted first_footer_offset in streaming messages                                                                                            
    - JSON serialization of reports and issue codes           


### Contributor Declaration

By opening this pull request, I affirm the following:

* All authors agree to the [Contributor License Agreement](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md).
* The code follows the project's coding standards.
* I have performed self-review and added comments where needed.
* I have added or updated tests to verify that my changes are effective and functional.
* I have run all existing tests and confirmed they pass.
 

<!-- PREVIEW-PUBLISH-TENSOGRAM-DOCS_BEGIN -->
Docs Preview
https://sites.ecmwf.int/docs/dev-section/tensogram/pull-requests/PR-22
<!-- PREVIEW-PUBLISH-TENSOGRAM-DOCS_END -->

### CODE REVIEW

File: docs/src/cli/validate.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+| canonical | 1–3 | Default + RFC 8949 CBOR key ordering |
+
+The mode flags (`--quick`, `--checksum`, `--canonical`) are mutually exclusive.
+
+## Options
+
+| Flag | Description |
+|------|-------------|
+| `--json` | Machine-parseable JSON output |
+
```
[the remark this answers is not in the document]
Comment (sametd) [reply docs/src/cli/validate.md:37]: The source files use standard single-pipe | syntax. This appears to be a rendering artifact in your analysis — there are no double leading pipes in our markdown. No change needed.

File: crates/tensogram-cli/src/commands/validate.rs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            .count();
+        let file_issue_count = report.file_issues.len();
+        let total_problems = msg_errors + file_issue_count;
+        eprintln!(
+            "{}: FAILED ({} errors, {} messages, {} objects)",
+            path.display(),
+            total_problems,
+            msg_count,
+            obj_count,
+        );
```
[the remark this answers is not in the document]
Comment (sametd) [reply crates/tensogram-cli/src/commands/validate.rs:133]: By design. File-level issues (gaps, trailing bytes) cause validation failure (is_ok() returns false, exit code 1). They are printed with a WARNING prefix on individual lines because they aren't message-level errors, but they are problems that fail the file. The summary correctly counts them as errors. See the doc comment on FileValidationReport::is_ok() which explains this. No change needed.

File: crates/tensogram-core/src/validate/metadata.rs [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+        }
+        let known_compressions = ["none", "szip", "zstd", "lz4", "blosc2", "zfp", "sz3"];
+        if !known_compressions.contains(&desc.compression.as_str()) {
+            issues.push(err(
+                IssueCode::UnknownCompression,
+                ValidationLevel::Metadata,
+                Some(i),
+                None,
+                format!("unknown compression '{}'", desc.compression),
+            ));
```
[the remark this answers is not in the document]
Comment (sametd) [reply crates/tensogram-core/src/validate/metadata.rs:303]: By design. Level 2 validates format correctness — a message with zfp compression is valid wire format regardless of whether this build has the zfp feature enabled. Level 3 catches unsupported codecs at decode time via PipelineConfigFailed. Added a comment documenting this. No change needed.

File: crates/tensogram-cli/src/main.rs [hunk 4]
```diff
@@ -120,6 +120,24 @@ enum Commands {
...
+        /// Quick mode: structure only (level 1)
+        #[arg(long, group = "vmode")]
+        quick: bool,
+        /// Checksum only: hash verification (level 3)
+        #[arg(long, group = "vmode")]
+        checksum: bool,
+        /// Canonical mode: levels 1-3 plus canonical CBOR check
+        #[arg(long, group = "vmode")]
+        canonical: bool,
+        /// Machine-parseable JSON output
```
[the remark this answers is not in the document]
Comment (sametd) [reply crates/tensogram-cli/src/main.rs:137]: This is incorrect. Clap's #[arg(group = "vmode")] on bool flags with the same group name does enforce mutual exclusivity by default. Verified empirically:                                                                                                                                     
                                                                                                                                             
$ tensogram validate --quick --checksum foo.tgm               
error: the argument '--quick' cannot be used with '--checksum'                                                                                   

This has been tested and confirmed across multiple review rounds. No change needed.

**Comment by tlmquintino** [comment]
PR #22 Analysis: feature/validate-cli

  Scope: 2727 lines, 18 files, 33 commits (10 "Copilot review" fix rounds, which I'll comment on).

  Architecture — well-structured

  The validate module is cleanly split into 4 files under crates/tensogram-core/src/validate/:
  - types.rs (235 lines) — Public types, ~40 IssueCode variants, serde serialization as snake_case strings. ValidationReport, FileValidationReport. Clean helper functions err(), warn().
  - structure.rs (651 lines) — Level 1 raw byte walker. Overflow-safe arithmetic throughout (checked_add, checked_mul, usize::try_from). Returns Option<FrameWalkResult> — None =
  unrecoverable parse failure, Some = walk completed (may have partial issues).
  - metadata.rs (344 lines) — Level 2 CBOR validation. Checks global metadata, index offsets, hash frame, preceder metadata, per-object descriptors (ndim/shape/strides consistency, known
  encoding/filter/compression, _reserved_.tensor).
  - integrity.rs (201 lines) — Level 3 hash verification with check_hash() helper that distinguishes Verified/Skipped/Failed. Runs decode_pipeline() for compression/encoding verification.
   hash_verified is true only when all objects are actually verified.
  - mod.rs (758 lines) — Public API entry points + 31 tests.

  Key design decisions — sound

  1. Streaming file validation: validate_file() uses scan_file() + per-message reads = O(1 message) memory.
  2. CBOR-before boundary: Uses Cursor::position() to measure consumed CBOR bytes (not reserialization). This is the correct approach — reserialization of non-canonical CBOR would produce
   different lengths, silently shifting the payload boundary.
  3. Mode dispatch: Structure always runs (needed to extract frames). Checksum mode suppresses structural warnings but keeps errors — correct tradeoff.
  4. hash_verified forced false on any error (mod.rs:76-78) — critical safety net.
  5. Missing hashes → warning not error — matches wire format spec (hashes are optional).

  framing.rs change — correct improvement

  The CBOR-before path in decode_data_object_frame() previously re-serialized CBOR to find the boundary. Now uses Cursor::position() + .deserialized() on the parsed ciborium::Value. This
  is:
  - More correct (non-canonical CBOR preserved)
  - More efficient (one parse instead of parse + reserialize)
  - Backwards-compatible (same function signature, same behavior for valid messages)

  Test coverage — adequate (41 tests total)

  - 31 tests in tensogram-core validate module covering all levels, all modes, streaming, zero-object, corruption, edge cases
  - 10 tests in CLI covering valid files, all modes, JSON output, batch, missing file, zero files
  - CI passes all 670 workspace tests on both Linux and macOS

  Panics in production code — zero

  All unwrap() calls are inside #[cfg(test)] modules. Production code uses Result, try_from, checked_add/mul, and unwrap_or with safe fallbacks throughout.

  Concerns

  1. 33 commits with 10 "fix: address Copilot review round N" — suggests iterative AI-assisted development. The code quality is good despite this; the rounds clearly addressed real issues
   (overflow checks, cursor position, CBOR-before boundary, hash_verified forcing). Not a blocker, just unusual commit history.
  2. PR claims 52 tests, I count 41 (31 + 10). The discrepancy may be from earlier counting methodology or deleted tests during review rounds. Not a blocker — coverage is adequate.
  3. No Python/C++/FFI tests — explicitly out of scope (PR 3 of 3 per the description). Acceptable for a PR 1.
  4. structure.rs:596 — first_footer_pos.unwrap_or(msg_end) is safe but worth a comment explaining why msg_end is the correct fallback (no footers = footer section starts at postamble).
  Minor.
  5. Level 3 decompression allocates full decoded buffer (integrity.rs:143-144 comment acknowledges this). For very large objects, a discard/sink mode would be better. Fine as-is for
  validation use case, and the comment documents the limitation.

**Verdict: OK to merge**

The code is well-structured, handles edge cases carefully, has zero panics in production paths, and passes all CI checks. The 3-level architecture is clean and extensible for the
  planned Level 4 (PR 2). The framing.rs fix is a genuine correctness improvement. Test coverage is adequate for a PR 1 of 3.

**No blocking issues. Ready to merge.**
